server/src/main/java/org/elasticsearch/index/engine/LazySoftDeletesDirectoryReaderWrapper.java

```diff
@@ -260,4 +281,32 @@ public boolean initialized() {
             return materializedBits != null;
         }
     }
+
+    static int applySoftDeletes(DocIdSetIterator iterator, FixedBitSet bits) throws IOException {
+        assert iterator != null;
+        int newDeletes = 0;
+        int docID;
+
+        while ((docID = iterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+            if (bits.get(docID)) { // doc is live - clear it
+                bits.clear(docID);
+                newDeletes++;
+                // now that we know we deleted it and we fully control the hard deletes we can do correct
+                // accounting
+                // below.
+            }
+        }
+        return newDeletes;
+    }
+
+    static CacheKey createCacheKey(DirectoryReader in) {
+        try (DirectoryReader clone = DirectoryReader.open(in.directory())) {
```

Comment:
This feels like too much of a hack to bring into ES. It's effectively opening up the index multiple times now whenever a search hits a shard in the frozen tier (one extra time per lucene index, and then once per segment in the Lucene index) . Let's rather work towards exposing the right abstraction in Lucene.

Reply:
Yeah I agree, it's a total hack. I couldn't figure out a better way to get to a unique CacheKey. Let's try to get a change in Lucene.